### PR TITLE
The AdminMenus created from Adminmenu-Module have higher selectionpriority than the other menu

### PR DESCRIPTION
**Pull request by stantoxt**
Admin Menus created from AdminMenuModule which have higher selectionPriority are not selected.
What I have done:
1.AutoAppend isAdminMenuNode=true QueryString For MenuItem craeted from AdminMenus
2.Check if MenuItem Href contains this querystring to match the selected result in the method MarkAsSelectedIfMatchesRouteOrUrl.

### CODE REVIEW

**Comment by Skrypt** [comment]
Hmm, I'm not sure about this one as we introduced also a Position param.

**Comment by stantoxt** [comment]
@Skrypt Can you give me some code?I can not understand it.

**Comment by Skrypt** [comment]
Share me a repro steps or some screenshot please.

**Comment by Skrypt** [comment]
Ok from TheBlogTheme we have an Admin Menu created from a recipe.
This menu has an menu item called "Blog" which points to `Admin/Contents/ContentItems/4erwd4sa9bdqssfdk2cpx2e0xs/Display`.

I created a second Admin Menu with a menu item called "Blog". So they should conflict with each others and they should take one or the other. So, if I'm not setting any priority on them it will pick the Menu Item from the recipe created one.

Though, if I set the priority to 1 on my manually created menu item it will take precedence over the recipe created one.

Also, if I do the opposite and set the priority to -1 on the recipe one. The manually created one will show.

So, this is working properly, though maybe I didn't get what you meant and there's a different use case in your scenario. Let me know.

**Comment by stantoxt** [comment]
Step1. This is the default Blog click event,the content/contentitems menu has a higher priority.
defaultblogclick
Step2. I wish blog menu could have a higher priority,after I change its priority,but actually it works like the step1.
defaultblogclickaftersetpriority

**Comment by Skrypt** [comment]
I'm sorry, I did not understand the issue. You mean that the Blog menu item doesn't get highlighted as the selected menu option when clicking on it because it doesn't find a matching url when passing in the method MarkAsSelectedIfMatchesRouteOrUrl. Ok, now I see your point. I understand now the issue, though I'm still not sure about the fix. Let me see if we can fix this without passing an extra url param and I'm not sure why your are using item.SelectionPriority instead of item.Priority.

**Comment by Skrypt** [comment]
Ok a change has been made in the MarkAsSelectedIfMatchesRouteOrUrl recently by @jtkech.

```C#
//from
match = viewContext.HttpContext.Request.Path.Equals(path, StringComparison.OrdinalIgnoreCase);

//to
match = viewContext.HttpContext.Request.Path.StartsWithSegments(path, StringComparison.OrdinalIgnoreCase);
```

This is what is creating the issue ; as multiple url's could start with `Admin/Contents/ContentItems/` as seen im the next screenshot.

image

So here it will make the last content item found in the Admin Menu as the selected one ; which is in that case the "Content Items" menu.

Concerning your PR. Even here the selection priority param would not make any sense since it would be applied on all items that comes from an admin menu which doesn't make a distinction between them.

**Comment by Skrypt** [comment]
See https://github.com/OrchardCMS/OrchardCore/issues/3431#issuecomment-526786705 and
https://github.com/OrchardCMS/OrchardCore/pull/4132

**Comment by Skrypt** [comment]
Here is what I did that works for me now : 

```C#

        private static void MarkAsSelectedIfMatchesRouteOrUrl(MenuItem menuItem, dynamic menuItemShape, ViewContext viewContext)
        {
            // compare route values (if any) first
            bool match = menuItem.RouteValues != null && RouteMatches(menuItem.RouteValues, viewContext.RouteData.Values);

            // if route match failed, try comparing URL strings, if
            if (!match && !String.IsNullOrWhiteSpace(menuItem.Href) && menuItem.Href[0] == '/')
            {
                PathString path = menuItem.Href.TrimEnd('/');
                PathString viewContextPath = viewContext.HttpContext.Request.Path.Value.TrimEnd('/');

                if (viewContext.HttpContext.Request.PathBase.HasValue)
                {
                    // Need to evaluate if this is still appropriate
                    if (path.StartsWithSegments(viewContext.HttpContext.Request.PathBase, StringComparison.OrdinalIgnoreCase, out var remaining))
                    {
                        path = remaining;
                    }
                }

                match = viewContextPath.Equals(path, StringComparison.OrdinalIgnoreCase);
            }

            menuItemShape.Selected = match;
        }

```

**Comment by Skrypt** [comment]
Ok so when we are executing `MarkAsSelectedIfMatchesRouteOrUrl(MenuItem menuItem, dynamic menuItemShape, ViewContext viewContext)` there is no indication on the `MenuItem` if this is a menu item that is coming from the **Admin Menus** module. So, what I don't like with this PR is that you are using the ViewContext instead to evaluate that a menu item is from the **Admin Menus**. `MenuItem` is totally ignorant about the fact that it's a `LinkAdminNode` or `PlaceholderAdminNode`. The only distinction between those is that one has a URL and the other doesn't. So it makes sense that you have used this technique but that's a hack TBH.

Here, the idea I have is to keep the StartsWithSegments() when we are parsing a MenuItem that is built for all url's that are not coming from the **Admin Menus** module.

One thing I noticed. StartsWithSegment will not work for example with the "Templates" routes :  

```
https://localhost:44300/OrchardCore.Templates/Template/Index
https://localhost:44300/OrchardCore.Templates/Template/Create
```

These 2 urls are returning false as they are not "starting" with the same segments but they are coming from the same TemplateController.cs. So, using the current technique will not cover 100% of scenarios. Right now, the way I see this is that in some cases we should parse the RouteValues and evaluate the Controller name only and in some other cases we need to parse the Url too.

File: src/OrchardCore/OrchardCore.Navigation.Core/NavigationHelper.cs [hunk 1]
```diff
@@ -176,7 +191,7 @@ private static dynamic GetHighestPrioritySelectedMenuItem(dynamic parentShape)
                     }
                     else // found more selected: tie break required.
                     {
-                        if (item.Priority > result.Priority)
+                        if (item.SelectionPriority > result.SelectionPriority)
```
Comment (stantoxt) [inline src/OrchardCore/OrchardCore.Navigation.Core/NavigationHelper.cs]: menuItemShape is not set up with Priority, it should use SelectionPriority Or Item.Priority

**Comment by stantoxt** [comment]
> ```cs
> ```cs
>             }
> 
>             menuItemShape.Selected = match;
>         }
> ```
> 
> 
> Do not take as a definite solution. @jtkech Skype me when you can. We need to find a solution that will handle both scenarios. Maybe treating the Admin Menus items differently with a higher SelectionPriority would make sense but I'm not convinced yet.
> ```



> See [#3431 (comment)](https://github.com/OrchardCMS/OrchardCore/issues/3431#issuecomment-526786705) and
> #4132
> 
> Ok so the intent of @jtkech in this PR is to allow selecting/highlighting a menu item while editing or adding content items. This is why he's doing a string.StartsWithSegments() comparison. Though, as we can see it introduces a regression on "Admin Menus" items. We need to find a better way than this to evaluate if we should select a menu item or not. Only using StartsWithSegments here is not accurate enough.

thanks @Skrypt ,This is great info for help me understand this.
> menuItemShape is not set up with Property Priority, it should use item.SelectionPriority Or item.Item.Priority
Maybe Here is a bug. This PR can fix the AdminMenu Nodes selectionPriority after setting up these menus priotiry.

**Comment by Skrypt** [comment]
The SelectionPriority vs Priority thing is strange as I tested and it seemed to work fine with only Priority. Also when we build the menu we set the SelectionPriority of the shape with the Priority of the MenuItem. Though it would make sense to use the SelectionPriority instead of Priority where you changed it because we are looking at a shape in that case.

**Comment by stantoxt** [comment]
> The SelectionPriority vs Priority thing is strange as I tested and it seemed to work fine with only Priority. 

Can you share me your repos or screenshot?

**Comment by jtkech** [comment]
After a quick look, what i saw is that in `NavigationHelper.BuildMenuItemShapeAsync`

                .SelectionPriority(menuItem.Priority)

is useless, maybe a typo, after replacing it with

                .Priority(menuItem.Priority)

setting a priority > 0 on the blog menu item was working.

Yes, what i did is incomplete, as i remember it was just relying on a simple url convention by only checking the starting segments, knowing that this convention is not respected many times.

At least, fixing the Priority allows to still select some menu items.

Maybe one solution would be to revert my change in `MarkAsSelectedIfMatchesRouteOrUrl` and then, if no munu item is selected, to be able to just keep selected the previous one (that was selected).

At least easy to say ;)

**Comment by Skrypt** [comment]
This can be fixed if we are using an IMenuItem I think... that way we will be able to know which type is the MenuItem coming from and evaluate url's properly according so.

**Comment by stantoxt** [comment]
> is useless, maybe a typo, after replacing it with
> 
> ```
>             .Priority(menuItem.Priority)
> ```
> 
> setting a priority > 0 on the blog menu item was working.
> 
I think maybe the typo error is here NavigationHelper.GetHighestPrioritySelectedMenuItem where we want to compare selectedMenuItem SelectionPriorityPriority
 ```
           if (item.Priority > result.Priority)
```
to
```
           if (item.SelectionPriority > result.SelectionPriority)
```

**Comment by stantoxt** [comment]
> This can be fixed if we are using an IMenuItem I think... that way we will be able to know which type is the MenuItem coming from and evaluate url's properly according so.

I hope we can distinguish between the menu from development and menu from admin dashboard.

**Comment by jtkech** [comment]
> I think maybe the typo error is here NavigationHelper.GetHighestPrioritySelectedMenuItem

Yes, because here we are cheking shapes, so if we keep `SelectionPriority` when initializing the shape, it would work the same, so for me your fix ( at least about the priority) look right ! Then about improving the menu selection more globally is another thing.

**Comment by stantoxt** [comment]
That's greate.I want to contribute on this.

**Comment by jtkech** [comment]
> I hope we can distinguish between the menu from development and menu from admin dashboard.

Hmm, not sure of the benefit here, admin menus / nodes use somewhere the same navigation core services that another menu provider would use, e.g `BuildMenuItemShapeAsync` and `GetHighestPrioritySelectedMenuItem` are in `NavigationHelper` belonging to `OrchardCore.Navigation.Core`. And the same shapes will be used, currently defined by the `OC.Navigation` module.

**Comment by jtkech** [comment]
@stantoxt i'm working on it and when i will have something to show i will do a 1st PR. Then feel free to open your own PR, as you want to contribute on this, and then use or not my code ;)

The idea i'm working on is to have a `Score` property, a total match giving an higher score than just `StartsWithSegments`, maybe check the number of matching segments and so on.

Note: With what i'm doing, if you go through the content items page to edit the blog, on the left the selected menu item becomes the `Blog` one because it exactly matches the current path.

Do you think that's ok? Knowing that with the priority fix we can do some more tweaks.

**Comment by stantoxt** [comment]
@jtkech That's wonderful.

**Pull request closed without merging**